File: netlify/functions/download.js
const axios = require('axios');
const { exec } = require('child_process');
const fs = require('fs');
const path = require('path');
const os = require('os');
const util = require('util');

const execPromise = util.promisify(exec);

exports.handler = async function(event, context) {
  // Only allow POST
  if (event.httpMethod !== 'POST') {
    return {
      statusCode: 405,
      body: JSON.stringify({ error: 'Method not allowed' })
    };
  }

  try {
    const { url, quality = 320 } = JSON.parse(event.body);

    if (!url) {
      return {
        statusCode: 400,
        body: JSON.stringify({ error: 'Please provide a YouTube URL' })
      };
    }

    if (!url.includes('youtube.com') && !url.includes('youtu.be')) {
      return {
        statusCode: 400,
        body: JSON.stringify({ error: 'Please provide a valid YouTube URL' })
      };
    }

    // Create a temporary directory
    const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'yt-'));
    const outputTemplate = path.join(tempDir, '%(title)s.%(ext)s');

    // Download the video using yt-dlp
    const command = `yt-dlp -f "best[height<=${quality}]" -o "${outputTemplate}" --no-playlist "${url}"`;
    
    try {
      await execPromise(command);
    } catch (error) {
      return {
        statusCode: 500,
        body: JSON.stringify({ error: `Error downloading video: ${error.message}` })
      };
    }

    // Find the downloaded file
    const files = fs.readdirSync(tempDir);
    
    if (files.length === 0) {
      return {
        statusCode: 500,
        body: JSON.stringify({ error: 'Download completed but no file was found' })
      };
    }

    const filePath = path.join(tempDir, files[0]);
    const fileContent = fs.readFileSync(filePath);
    
    // Clean up
    fs.unlinkSync(filePath);
    fs.rmdirSync(tempDir);

    return {
      statusCode: 200,
      headers: {
        'Content-Type': 'video/mp4',
        'Content-Disposition': `attachment; filename="${files[0]}"`
      },
      body: fileContent.toString('base64'),
      isBase64Encoded: true
    };

  } catch (error) {
    return {
      statusCode: 500,
      body: JSON.stringify({ error: `An unexpected error occurred: ${error.message}` })
    };
  }
}; 

File: netlify/functions/download.py
import json
import os
import subprocess
import uuid
import tempfile
from http.server import BaseHTTPRequestHandler

def download_video(url, quality=320):
    """Download a YouTube video at specified quality using yt-dlp."""
    try:
        # Create a temporary directory for the download
        temp_dir = tempfile.mkdtemp()
        output_template = f"{temp_dir}/%(title)s.%(ext)s"
        
        # Format the command to download the video at specified quality
        command = [
            "yt-dlp",
            "-f", f"best[height<={quality}]",
            "-o", output_template,
            "--no-playlist",
            url
        ]
        
        # Execute the command
        result = subprocess.run(command, capture_output=True, text=True, check=True)
        
        # Find the downloaded file
        files = os.listdir(temp_dir)
        
        if not files:
            return {
                "statusCode": 500,
                "body": json.dumps({"error": "Download completed but no file was found"})
            }
        
        # Get the file path
        file_path = os.path.join(temp_dir, files[0])
        
        # Read the file content
        with open(file_path, 'rb') as f:
            file_content = f.read()
        
        # Clean up
        os.remove(file_path)
        os.rmdir(temp_dir)
        
        return {
            "statusCode": 200,
            "headers": {
                "Content-Type": "video/mp4",
                "Content-Disposition": f"attachment; filename={files[0]}"
            },
            "body": file_content,
            "isBase64Encoded": True
        }
        
    except subprocess.CalledProcessError as e:
        return {
            "statusCode": 500,
            "body": json.dumps({"error": f"Error downloading video: {e.stderr}"})
        }
    except Exception as e:
        return {
            "statusCode": 500,
            "body": json.dumps({"error": f"An unexpected error occurred: {str(e)}"})
        }

def handler(event, context):
    """Netlify serverless function handler."""
    if event["httpMethod"] != "POST":
        return {
            "statusCode": 405,
            "body": json.dumps({"error": "Method not allowed"})
        }
    
    try:
        body = json.loads(event["body"])
        url = body.get("url", "").strip()
        quality = int(body.get("quality", 320))
        
        if not url:
            return {
                "statusCode": 400,
                "body": json.dumps({"error": "Please provide a YouTube URL"})
            }
        
        if not ("youtube.com" in url or "youtu.be" in url):
            return {
                "statusCode": 400,
                "body": json.dumps({"error": "Please provide a valid YouTube URL"})
            }
        
        return download_video(url, quality)
        
    except json.JSONDecodeError:
        return {
            "statusCode": 400,
            "body": json.dumps({"error": "Invalid JSON in request body"})
        }
    except Exception as e:
        return {
            "statusCode": 500,
            "body": json.dumps({"error": f"An unexpected error occurred: {str(e)}"})
        } 